sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala

```diff
@@ -83,6 +116,20 @@ object SubqueryExpression {
       case _ => false
     }.isDefined
   }
+
+  /**
+   * Clean the outer references by normalizing them to BindReference in the same way
+   * we clean up the arguments during LogicalPlan.sameResult. This enables to compare two
+   * plans which has subquery expressions. This method returns a [[CanonicalizedSubqueryExpr]]
+   * which wraps the underlying [[SubqueryExpression]].
+   */
+  def canonicalize(e: SubqueryExpression, attrs: AttributeSeq): CanonicalizedSubqueryExpr = {
+    // Normalize the outer references in the subquery plan.
+    val subPlan = e.plan.transformAllExpressions {
+      case OuterReference(r) => BindReferences.bindReference(r, attrs, allowFailures = true)
```

Comment:
@cloud-fan Hi Wenchen, If we are able to normalize all the expression ids, then i think we can solve the nondeterministic problem. I had actually tried this for the new test cases i have added. But i didn't go with this as i felt i may have to test this more.  In my testing, i only normalized the AttributeReference .. but we can have generated expression ids for other expressions, right , Aliases, SubqueryExpression etc. We also need to normalize these as well, no ?